### PR TITLE
Test encodeKey and order invariants

### LINKED ISSUE #4464: Key encoding for store messing up for -0
## Describe the bug
If you try to use -0 as a key into a BigMap or one of the other store types, it encodes as NaN

## Expected behavior
It should encode as 0.

Possibly there may also be issues with NaN itself as well as the Infinities.  These should get checked out as part of this, and the relevant test (`test-collections.js`) should test all the weird IEEE float values.

### PR DESCRIPTION
**Pull request by erights**
Fixed #4464 
On the encoding of -0. Instead it now canonicalizes -0 to 0, reflecting exactly the semantics of our distributed object systems (as well as JSON.stringify).

Now canonicalizes -0 to 0 on decoding too.

Now canonicalizes the encoded form of NaN.

### CODE REVIEW

**Review by FUDCo — CHANGES_REQUESTED** [review]
Mostly good, but I'd be happier if the test included an explicit test of the Infinities.  I'm pretty confident they'll round trip through the encoding just fine but I'm not 100% sure they'll sort properly.

**Comment by erights** [comment]
> Mostly good, but I'd be happier if the test included an explicit test of the Infinities. I'm pretty confident they'll round trip through the encoding just fine but I'm not 100% sure they'll sort properly.

Done. Ready for review. PTAL

**Comment by mhofman** [comment]
I added a test of the "negative" `NaN`, which would have broken before canonicalization.

As a matter of process, I'd prefer if refactoring and changes were in separate commits. I'm tempted to perform surgery on this branch to split the commits, to more easily track the changes.

**Comment by erights** [comment]
> I added a test of the "negative" `NaN`, which would have broken before canonicalization.
> 
> As a matter of process, I'd prefer if refactoring and changes were in separate commits. I'm tempted to perform surgery on this branch to split the commits, to more easily track the changes.

If you'd like to do that, by all means! Thanks!

**Comment by mhofman** [comment]
Surgery performed. Now the PR can be reviewed by commit.

I squashed my "negative" `NaN` test addition into the last commit, but for reference, here it was standalone: https://github.com/Agoric/agoric-sdk/commit/240abf1ad6de5f542de6614b00dcedac2bb9e1e2

File: packages/store/src/patterns/encodeKey.js [hunk 1]
```diff
@@ -38,7 +40,17 @@ const asBits = new BigUint64Array(asNumber.buffer);
...
 
+// TODO Choose the same canonical NaN encoding that cosmWasm and ewasm chose.
+const CanonicalNaN = 'ffff8000000000000';
+
+// Normalize -0 to 0
+
 const numberToDBEntryKey = n => {
+  if (is(n, -0)) {
+    n = 0;
+  } else if (is(n, NaN)) {
```
Comment (mhofman) [inline packages/store/src/patterns/encodeKey.js:51]: Out of curiosity, why not use `Number.isNaN`?

I rarely remember the equality rules for `NaN`.

File: packages/store/src/patterns/encodeKey.js [hunk 2]
```diff
@@ -62,7 +74,11 @@ const dbEntryKeyToNumber = k => {
     bits ^= 0x8000000000000000n;
   }
   asBits[0] = bits;
-  return asNumber[0];
+  const result = asNumber[0];
+  if (is(result, -0)) {
+    return 0;
+  }
+  return result;
```
Comment (mhofman) [inline packages/store/src/patterns/encodeKey.js:81]: Under what circumstances would a negative zero end up as a key in the store? Or is it just defense in depth?

File: packages/store/test/test-encodeKey.js [hunk 3]
```diff
@@ -40,12 +40,27 @@ const decodeKey = makeDecodeKey(decodeRemotable);
...
+
+const NegativeNaN = getNaN('ffffffffffffffff');
+
 const goldenPairs = harden([
   [37n, 'p0000000002:37'],
   [-1n, 'n9999999999:9'],
   [1, 'fbff0000000000000'],
   [-1, 'f400fffffffffffff'],
   [NaN, 'ffff8000000000000'],
+  [NegativeNaN, 'ffff8000000000000'],
```
Comment (mhofman) [inline packages/store/test/test-encodeKey.js:63]: Without canonicalization of `NaN` this would have ended up as `0` when decoding back.

**Review by mhofman — APPROVED** [review]

File: packages/store/src/patterns/encodeKey.js (hunk 1 above)
Comment (erights) [reply packages/store/src/patterns/encodeKey.js:51]: No reason except that I was already using `is`.

File: packages/store/src/patterns/encodeKey.js (hunk 2 above)
Comment (erights) [reply packages/store/src/patterns/encodeKey.js:81]: Just defense in depth.

File: packages/store/test/test-encodeKey.js (hunk 3 above)
Comment (erights) [reply packages/store/test/test-encodeKey.js:63]: Cool! I just barely fixed it before you attacked ;)

**Comment by erights** [comment]
@FUDCo PTAL. Because you had requested changes, I'm blocked on your approval. Thanks.

**Review by FUDCo — APPROVED** [review]
LGTM

**Comment by mhofman** [comment]
@erights given the refactoring, it would have been preferable to merge this as an actual merge, not squashing.

**Pull request merged**